lib/index.js

```diff
@@ -12,34 +26,307 @@ class PluginPassportOAuth {
+            if (!this.config.strategies[key].credentials) {
+              return reject(new this.context.errors.BadRequestError(`Error loading strategy [${key}]: no credentials provided`));
+            }
+
+            try {
+              this.strategyConstructor = require('passport-' + key).Strategy;
+            } catch (err) {
+              return reject(new this.context.errors.BadRequestError(`Error loading strategy [${key}]: ${err.message}`));
+            }
+
+            try {
+              refresh.use(this.strategyConstructor);
+            } catch (err) {
+              /*eslint no-console: 0*/
+              console.error(`Error refreshing strategy [${key}]: ${err.message}`);
+            }
+          });
+          this.initStrategies();
+          return resolve();
+        });
+      });
+  }
+
+  initStrategies () {
+    this.strategies = {};
+    Object.keys(this.config.strategies).forEach(strategy => {
+      this.strategies[strategy] = {
+        config: {
+          constructor: this.strategyConstructor,
+          strategyOptions: this.config.strategies[strategy].credentials,
+          authenticateOptions: {
+            scope: this.config.strategies[strategy].scope
+          },
+          fields: ['access_token', 'refresh_token'],
+          verify: 'verify'
+        },
+        methods: {
+          exists: 'exists',
+          validate: 'validate',
+          create: 'create',
+          update: 'update',
+          delete: 'delete',
+          getInfo: 'getInfo'
+        }
+      };
+    });
+  }
+
+  /**
+   * Get the repository according to the strategy
+   *
+   * @param provider
+   * @returns {module.exports.constructors.Repository|*}
+   */
+  getProviderRepository (provider) {
+    if (!this.providerRepository) {
+      this.providerRepository = new this.context.constructors.Repository(provider);
     }
 
-    strategy = new Strategy(this.context, this.config);
-    strategy.init();
+    return this.providerRepository;
+  }
+
+  /**
+   * Get the credentials of a given userId from the repository of the corresponding strategy
+   *
+   * @param userId
+   * @param provider
+   * @returns {Promise|Promise.<TResult>}
+   */
+  getCredentialsFromUserId (userId, provider) {
+    return this.getProviderRepository(provider).search({
+      query: {
+        match: {
+          userId
+        }
+      }
+    })
+      .then(result => {
+        if (result.total === 0) {
+          return Promise.resolve(null);
+        }
+
+        return Promise.resolve(result.hits[0]);
+      });
+  }
+
+  /**
+   * Verify function called by passport to log the user.
+   * If a persist option is in the configuration of the strategy then every field present in this configuration will be persisted in the repository.
+   *
+   * @param request
+   * @param accessToken
+   * @param refreshToken
+   * @param profile
+   * @returns {Promise.<T>}
+   */
+  verify (request, accessToken, refreshToken, profile) {
```

Comment:
Following implementation look better for me (can be discussed):

```javascript
verify (request, accessToken, refreshToken, profile) {
  return this.getCredentialsFromUserId(profile._json[this.config.strategies[profile.provider].identifierAttribute], profile.provider)
    .then(oauthUserObject => {
      if (oauthUserObject !== null) {
        return Promise.resolve(userObject._id); 
      }

      // Here we still have no credentials stored for current strategy with given user. It can be due to 2 cases:
      // 1. the user still not exists in Kuzzle => we can create it (see next step)
      // 2. the user exists but with no credentials for current strategy => we have no mean to know that automatically, so we have no other choice to ignore this case

      if (this.config.strategies[profile.provider].persist.length !== 0) { 
        const user = {};
        Object.keys(profile._json).forEach(attr => {
        if (this.config.strategies[profile.provider].persist.indexOf(attr) > -1) {
          user[attr] = profile._json[attr];
        }
        // Call [Security controller](http://docs.kuzzle.io/api-reference/?others#createuser) to create the user with current strategy's credentials, and get it's ID
        // (...)
       // (result = response result from security.createUser() request)
        return Promise.resolve(result_id); 
      }

     // Here we did not find any user mapped to given oauth credentials, (and did not create it, due to plugin options) => deny the authentication
      return Promise.resolve(null); 
    })
    .catch(err => Promise.reject(err)); 
  });
}
```